funtimes/R/wavk.test.R

```diff
@@ -1,192 +1,169 @@
-                a <- ar(TS, aic=FALSE, order.max=which.min(bic), demean=TRUE)
-                pheta <- a$ar
-            }
-        } else { #BIC==FALSE, use fixed ar.order>0
-            if (robust){
-                pheta <- HVK(TS, ar.order=ar.order)
-            } else {
-                a <- ar(TS, aic=FALSE, order.max=ar.order, demean=TRUE)
-                pheta <- a$ar
-            }
+      if (BIC) {
+        bic <- rep(NA, ar.order)
+        if (robust){
+          for (i in 1:length(bic)) {
+            pheta_rob <- HVK(TS, ar.order=i)
+            tmp <- filter(TS, pheta_rob, sides=1)
+            et <- TS[(i+1):n] - tmp[i:(n-1)]
+            bic[i] <- (n-i)*log(var(et)) + i*log(n-i)
+          }
+          pheta <- HVK(TS, ar.order=which.min(bic))
+        } else {
+          for (i in 1:length(bic)) {
+            a <- ar(TS, aic=FALSE, order.max=i, demean=TRUE)
+            bic[i] <- n*log(a$var.pred) + i*log(n)
+          }
+          a <- ar(TS, aic=FALSE, order.max=which.min(bic), demean=TRUE)
+          pheta <- a$ar
+        }
+      } else { #BIC==FALSE, use fixed ar.order>0
+        if (robust){
+          pheta <- HVK(TS, ar.order=ar.order)
+        } else {
+          a <- ar(TS, aic=FALSE, order.max=ar.order, demean=TRUE)
+          pheta <- a$ar
         }
-        names(pheta) <- paste(rep("phi_", length(pheta)), c(1:length(pheta)), sep="")
+      }
+      names(pheta) <- paste(rep("phi_", length(pheta)), c(1:length(pheta)), sep="")
     } else { #ar.order==0, no filtering
-        pheta <- numeric(0)
+      pheta <- numeric(0)
     }
-    
     if (length(pheta)>0) {
-        tmp <- filter(x, pheta, sides=1)
-        if (H0 == "linear") {
-            tmp2 <- filter(mod$fitted.values, pheta, sides=1)
-            Z <- (x[(length(pheta)+1):n] - tmp[length(pheta):(n-1)]) - (mod$fitted.values[(length(pheta)+1):n] - tmp2[length(pheta):(n-1)])
-            ESTIMATE <- list(linear, pheta)
-            names(ESTIMATE) <- c("linear_trend_coefficients", "AR_coefficients")
-        } else {
-            Z <- (x[(length(pheta)+1):n] - tmp[length(pheta):(n-1)])
-            ESTIMATE <- list(pheta)
-            names(ESTIMATE) <- c("AR_coefficients")
-        }
+      tmp <- filter(x, pheta, sides=1)
+      tmp2 <- filter(mod$fitted.values, pheta, sides=1)
+      Z <- (x[(length(pheta)+1):n] - tmp[length(pheta):(n-1)]) - (mod$fitted.values[(length(pheta)+1):n] - tmp2[length(pheta):(n-1)])
+      ESTIMATE <- list(linear, pheta)
+      names(ESTIMATE) <- c("trend_coefficients", "AR_coefficients")
     } else {
-        Z <- TS
-        if (H0 == "linear") {
-            ESTIMATE <- list(linear, pheta)
-            names(ESTIMATE) <- c("linear_trend_coefficients", "AR_coefficients")
-        } else {
-            ESTIMATE <- list(pheta)
-            names(ESTIMATE) <- c("AR_coefficients")
-        }
+      Z <- TS
+      ESTIMATE <- list(linear, pheta)
+      names(ESTIMATE) <- c("trend_coefficients", "AR_coefficients")
```

Comment:
lines 105-106 and 109-110 are now identical, and can be placed after the "if-else" loop, only once.